de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/negotiation/InstantOutgoingProjectNegotiation.java

```diff
@@ -0,0 +1,273 @@
+        if (transferList.isEmpty())
+            return;
+
+        log.debug(this + ": file transfer start");
+        assert fileTransferManager != null;
+
+        String message = "Sending files to " + getPeer().getName() + "...";
+        monitor.beginTask(message, transferList.size());
+
+        /* use piped stream to communicate with transfer thread */
+        PipedInputStream in = new PipedInputStream();
+        OutputStream out = new PipedOutputStream(in);
+
+        String userID = getPeer().toString();
+        OutgoingFileTransfer transfer;
+        transfer = fileTransferManager.createOutgoingFileTransfer(userID);
+
+        try {
+            OutgoingStreamProtocol osp;
+            osp = new OutgoingStreamProtocol(out, session, monitor);
+
+            /* id in description needed to bypass SendFileAction handler */
+            String streamName = TRANSFER_ID + getID();
+            transfer.sendStream(in, streamName, 0, streamName);
+
+            awaitNegotation(transfer, monitor);
+
+            sendProjectConfigFiles(osp);
+            sendRemainingPreferOpenedFirst(osp);
+
+            osp.close();
+        } finally {
+            IOUtils.closeQuietly(out);
+        }
+
+        monitor.done();
+        log.debug(this + ": file transfer done");
+    }
+
+    @Override
+    protected void cleanup(IProgressMonitor monitor) {
+        editorManager.removeSharedEditorListener(listener);
+        session.userStartedQueuing(remoteUser);
+
+        if (stoppedUsers != null)
+            startUsers(stoppedUsers);
+
+        super.cleanup(monitor);
+    }
+
+    private void createTransferList(List<FileList> fileLists, int fileCount) {
+        List<SPath> files = new ArrayList<SPath>(fileCount);
+        for (final FileList list : fileLists) {
+            IProject project = session.getProject(list.getProjectID());
+            for (String file : list.getPaths()) {
+                files.add(new SPath(project.getFile(file)));
+            }
+        }
+
+        /* sort hierarchy based, top files are seen first in project explorer */
+        Collections.sort(files, new Comparator<SPath>() {
+            @Override
+            public int compare(SPath a, SPath b) {
+                int lenA = a.getProjectRelativePath().segmentCount();
+                int lenB = b.getProjectRelativePath().segmentCount();
+                return Integer.valueOf(lenA).compareTo(Integer.valueOf(lenB));
+            }
+        });
+
+        /* LinkedHashSet for fast lookup while keeping sort order */
+        transferList = new LinkedHashSet<SPath>(files);
+    }
+
+    private void fileOpened(SPath file) {
+        if (file != null) {
+            openedFiles.addFirst(file);
+            log.debug(this + ": added " + file + " to open files queue");
+        }
+    }
+
+    private void sendProjectConfigFiles(OutgoingStreamProtocol osp)
+        throws IOException, LocalCancellationException {
+        /* TODO should be configurable in future */
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS1135))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=128&issues=AWSS35X1sk71k7lwr5eP&open=AWSS35X1sk71k7lwr5eP)

[ISSUE_KEY]: AWSS35X1sk71k7lwr5eP